README
NAME
    Perinci::Sub::Gen::Undoable - Generate undoable (transactional,
    dry-runnable, idempotent) function

VERSION
    version 0.25

SYNOPSIS
     # See an example, like in Setup::File::Symlink

DESCRIPTION
    This module helps you write undoable/transactional functions (as well as
    functions that support dry-run and are idempotent).

SEE ALSO
    Rinci::function::Undo, Rinci::function::Transaction, Rinci

    Setup.

DESCRIPTION
    This module has Rinci metadata.

FUNCTIONS
    None are exported by default, but they are exportable.

  gen_undoable_func(%args) -> [status, msg, result, meta]
    Generate undoable (transactional, dry-runnable, idempotent) function.

    This function (hereby called generator, or helper) is a helper for
    writing undoable (as well as transactional, dry-runnable, and
    idempotent) function (hereby called generated function). Undo and
    transaction features are as described in Rinci::function::Undo and
    Rinci::function::Transaction. Generator will create the generated
    function with the basic flow and structure; you just need to supply the
    necessary "business logics" bits like the "check_state" and "fix_state"
    hooks.

    Here is the flow of the generated function:

    1.  Execute "check_args" hook if supplied. Hook will be passed ($args).
        The task of the this hook is to check arguments (and possibly supply
        defaults). It should return enveloped result. If response is not
        success, exit with this response. Otherwise, continue.

    1.  If we are called in do mode (-undo_action => 'do' or undefined):

           2a. Execute C<check_state> hook, to check state. Hook will be passed ($args).
           Hook must return enveloped result, status 304 if checking succeeds and
           fixed state (final, desired state) has been reached so nothing needs to
           be done, status 200 if checking succeeds and we are in a fixable state (a
           non-final state but fixable into the final state, also called an
           acceptable initial state), status 412 if checking succeeds but we are in
           an unfixable state (non-final state which is not fixable into the final
           state, also called an unacceptable initial state). Other status means
           checking has failed. Function should return with this result unless
           status is 200.

           If status is 200, hook must also return undo data (a list of calls that
           can be used to I<undo> the 'fix_state' routine back to the initial state)
           in its result metadata, in C<undo_data> key. This means, if we fail during
           fixing, we can rollback to initial state using this information.

           The undo data looks like this: C. C
           is the fully-qualified function name to call (required), C is a
           hashref containing arguments for the function (optional, defaults to {}).

           2b. If function is called in dry run mode (-dry_run => 1), we return 200
           (along undo data from step 2a in the result metadata).

           2c. We execute C<fix_state> hook to fix state to final/fixed state. Hook will
           be passed ($args, $undoI<data), where $undo>data was retrieved from
           C<check_state> hook. Hook must return enveloped result. Function returns
           with this result. Failure will trigger a rollback by TM.

    1.  If we are called in undo mode (-undo*action =* 'undo', -undo>data =>
        $calls): then we simply invoke TM's call() to process the undo
        calls.

    For examples, see Setup::* Perl modules.

    Arguments ('*' denotes required arguments):

    *   args => *hash* (default: {})

        Specification for generated function's arguments.

        This is just like the metadata property 'args'.

    *   check_args => *code*

        Code to check function's arguments.

        See generator's documentation on how this hook should do and return.

    *   check_or_fix_state => *code*

        Code to check or fix state.

        An alternative to combine check*state and fix*state in one code.
        Code will be called with either ('check', $args) or ('fix', $args,
        $undo_data), depending on whether check/fix is wanted.

    *   check_state* => *code*

        Code to check state.

        See generator's documentation on how this hook should do and return.

    *   description => *str*

        Generated function's description.

    *   fix_state => *code*

        Code to fix state.

        See generator's documentation on how this hook should do and return.

    *   install => *bool* (default: 1)

        Whether to install generated function (and metadata).

        By default, generated function will be installed to the specified
        (or caller's) package, as well as its generated metadata into %SPEC.
        Set this argument to false to skip installing.

    *   name* => *str*

        Generated function's name, e.g. `myfunc`.

    *   package => *str*

        Generated function's package, e.g. `My::Package`.

        This is needed mostly for installing the function. You usually don't
        need to supply this if you set "install" to false.

        If not specified, caller's package will be used by default.

    *   summary => *str*

        Generated function's summary.

    *   trash_dir => *bool* (default: 0)

        Whether function needs undo trash directory.

        If set to true, generator will put "-undo_trash_dir" in function
        arguments.

    *   v => *int* (default: 1)

        Interface version.

        When Perinci::Sub::Gen::Undoable 0.22 broke backward-compatibility,
        v was introduced at v=2 (previous interface at v=1).

        For new interface, must be set to 2.

    Return value:

    Returns an enveloped result (an array). First element (status) is an
    integer containing HTTP status code (200 means OK, 4xx caller error, 5xx
    function error). Second element (msg) is a string containing error
    message, or 'OK' if status is 200. Third element (result) is optional,
    the actual result. Fourth element (meta) is called result metadata and
    is optional, a hash that contains extra information.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.

POD ERRORS
    Hey! The above document had some coding errors, which are explained
    below:

    Around line 341:
        You have '=item 1' instead of the expected '=item 2'

    Around line 375:
        You have '=item 1' instead of the expected '=item 3'



Changes
Revision history for Perinci-Sub-Gen-Undoable

0.25     2012-08-16 (SHARYANTO)

         [REMOVED FEATURES]

         - For the sake of simplicity and reducing duplicated code, rollback and
           undo/redo now depends solely on TM instead of this module
           reimplementing it. All functions generated by this function will
           require transaction.

         - Arguments 'use_tx' and 'req_tx' are also removed, consequently.


0.24     2012-08-16 (SHARYANTO)

         No functional changes. Require the appropriate version of
         Perinci::Tx::Manager. Forgot to delete old code
         (Perinci::Sub::Step::*).


0.23     2012-08-16 (SHARYANTO)

         - Add argument 'req_tx'.

         - Some minor tweaks/fixes.

         - Remove logging for sub entry/exit. Users should now use
           Log::Any::For::Package.


0.22     2012-08-15 (SHARYANTO)

         This release breaks backward compatibility. It introduce v and requires
         v=2 (previous version is at v=1). Arguments change (no more
         'build_steps', 'steps'), no more steps concept. Now follows Rinci
         1.1.23.


0.21     2012-08-10 (SHARYANTO)

         No functional changes. Use Perinci::Sub::Gen instead of the old
         Perinci::Sub::Gen::common (which caused circular dependency problem
         since it is contained in Perinci-Sub-Gen-AccessTable, which depends on
         Perinci-CmdLine, which depends on Perinci, which depends on us).


0.20     2012-08-10 (SHARYANTO)

         [ENHANCEMENTS]

         - In `steps` argument, instead of a stepspec hash, a string name
           containing name of another step can also be specified, for aliasing.


0.19     2012-08-10 (SHARYANTO)

         No functional changes. Suppress logging fix message when Common::call*
         is just doing check phase, using special argument -log_fix => 0.


0.18     2012-08-10 (SHARYANTO)

         [ENHANCEMENTS]

         - Get step specification from Perinci::Sub::Step::* if not specified in
           `steps` argument.

         - Provide Common::call and Common::call_undo steps.


0.17     2012-08-09 (SHARYANTO)

         Update to Rinci 1.1.22 (and Perinci::Tx::Manager 0.27), where we expect
         undo data to be in -undo_data argument instead of using
         get_undo_steps() and get_redo_steps(). This change is to handle nested
         calls.


0.16     2012-08-08 (SHARYANTO)

         - Use Perinci::Exporter.


0.15     2012-07-27 (SHARYANTO)

         [ENHANCEMENTS]

         - Step's fix_log_message can also be a coderef.


0.14     2012-07-27 (SHARYANTO)

         [ENHANCEMENTS]

         - Add argument: log_level_fix_step, default to 'info'.


0.13     2012-07-04 (SHARYANTO)

         [ENHANCEMENTS]

         - Pass result and result metadata to step's 'fix' routines to allow
           them to be set/changed.

         - Add `install` argument, if set to true then will install the function
           as well as the meta into target package.


0.12     2012-07-04 (SHARYANTO)

         Previous release new feature didn't work. Try again.


0.11     2012-07-04 (SHARYANTO)

         Allow unqualified name, will use caller's package to qualify.


0.10     2012-07-04 (SHARYANTO)

         Interpret 304 status (aside from 200) from step's check as success.


0.09     2012-07-04 (SHARYANTO)

         Sorry, another incompatible change. Hopefully no more in the near
         future.

         [INCOMPATIBLE CHANGES]

         - step's 'check' should return enveloped result, like step's 'fix'.
           This is to simplify code that uses 'check_or_fix'.


0.08     2012-06-26 (SHARYANTO)

         [ENHANCEMENTS]

         - step's 'check' hook can also be an arrayref, for convenience.

         - Add step's 'check_or_fix'.


0.07     2012-06-23 (SHARYANTO)

         There are a couple of incompatible changes. API is still in flux. Sorry
         about this.

         [INCOMPATIBLE CHANGES]

         - 'hook_check_args' renamed to 'check_args' and no longer documented as
           temporary.

         - step's hooks 'gen_undo' and 'run' renamed to 'check' and 'fix',
           respectively.

         [ENHANCEMENTS]

         - 'build_steps' can also be an array instead of code, for convenience.


0.06     2012-06-22 (SHARYANTO)

         [INCOMPATIBLE CHANGES]

         - 'build_steps' is now passed \%args instead of %args, to make it
           consistent with others. Sorry about this.


0.05     2012-06-21 (SHARYANTO)

         [FIXES]

         - Remove old code in 'redo' part.


0.04     2012-06-21 (SHARYANTO)

         [FIXES]

         - Fix setting function result during rollback.


0.03     2012-06-21 (SHARYANTO)

         [FIXES]

         - When calling record_call(), record original function arguments before
           being potentially modified by hook_check_args.

         - Grok if function is being called by txm during rollback (-tx_action
           is set to 'rollback'). Don't do our own rollback again under this
           condition.


0.02     2012-06-07 (SHARYANTO)

         Fix a couple of typos.


0.01     2012-06-06 (SHARYANTO)

         First release. Setup::File::Symlink works using this module. Minimal
         tests. The transactional part has not been tested (waiting for
         Perinci::Tx::Manager).


t/testlib.pl
use 5.010;
use strict;
use warnings;
use Log::Any '$log';

use Perinci::Sub::Gen::Undoable qw(gen_undoable_func);
use Test::More 0.96;

sub test_gen {
    my (%targs) = @_;

    subtest $targs{test_name} => sub {
        my $res;
        my %fargs = (
            v => 2,
            name               => $targs{func_name},
            summary            => $targs{summary},
            description        => $targs{description},
            check_state        => $targs{check_state},
            fix_state          => $targs{fix_state},
            check_or_fix_state => $targs{check_or_fix_state},
        );
        if ($targs{other_args}) {
            while (my ($k, $v) = each %{$targs{other_args}}) {
                $fargs{$k} = $v;
            }
        }
        eval { $res = gen_undoable_func(%fargs) };
        my $eval_err = $@;
        diag "died during function: $eval_err" if $eval_err;

        if ($targs{dies}) {
            ok($eval_err, "dies");
            return;
        } else {
            ok(!$eval_err, "doesn't die") or do {
                diag "dies: $eval_err";
                return;
            };
        }

        $targs{status} //= 200;
        is($res->[0], $targs{status}, "status = $targs{status}") or
            do { diag explain $res; return };

        if ($res->[0] == 200) {
            my $func = $res->[2]{code};
            my $meta = $res->[2]{meta};
            is(ref($func), 'CODE', 'func returned');
            is(ref($meta), 'HASH', 'meta returned');
        }

        if ($targs{post_test}) {
            $targs{post_test}->($res);
        }
    };
}

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601",
    "Test::More" => "0.96"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Generate undoable (transactional, dry-runnable, idempotent) function",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Perinci-Sub-Gen-Undoable",
  "dist_version" => "0.25",
  "license" => "perl",
  "module_name" => "Perinci::Sub::Gen::Undoable",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Log::Any" => 0,
    "Perinci::Exporter" => 0,
    "Perinci::Sub::Gen" => 0,
    "Perinci::Sub::Wrapper" => "0.29",
    "Perinci::Tx::Manager" => "0.29",
    "Rinci" => "v1.1.23",
    "SHARYANTO::Log::Util" => 0,
    "SHARYANTO::Package::Util" => 0,
    "perl" => "5.010001"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Perinci/Sub/Step/Common/call.pm
package Perinci::Sub::Step::Common::call;

use 5.010;
use strict;
use warnings;

use Perinci::Util qw(get_package_meta_accessor);

our $VERSION = '0.23'; # VERSION

# provide 'check_or_fix' functionality for several steps: call, call_undo,
# call_riap, call_riap_undo.
sub __check_or_fix_for_call_or_call_undo {
    # whichs = which step ('call', 'call_undo', 'call_riap', 'call_riap_undo')
    # whicha = which step action ('check' or 'fix')
    # cargs  = calling function's arguments
    my ($whichs, $whicha, $cargs, $step, $step_undo) = @_;

    # undo step name
    my $us;
    if ($whichs eq 'call') {
        $us = 'Common::call_undo';
    } elsif ($whichs eq 'call_undo') {
        $us = 'Common::call';
    } else {
        die "Sorry, $whichs is not supported yet";
    }

    # XXX to support call_riap and call_riap_undo, we need to provide the Riap
    # equivalent for getting metadata and for calling the function under
    # transaction. this support has not been added, maybe later when needed.

    # for 'call' and 'call_undo', f will be a fully-qualified function name,
    # e.g. Foo::Bar::func. for 'call_riap' and 'call_riap_undo', f will be a
    # Riap URL.
    my $f         = $step->[1];
    my $fargs0    = $step->[2];
    my $undo_data = $step->[3];

    # check step arguments
    if ($whicha eq 'check') {
        my $pkg = $f; $pkg =~ s/::\w+\z// or return [
            400, "Invalid function name (not fully-qualified): $f"];
        defined &{$f} or return [400, "Function does not exist: $f"];
        if (defined $fargs0) {
            ref($fargs0) eq 'HASH'
                or return [400, "Function arguments must be hash"];
        }
        my $res = get_package_meta_accessor(package=>$pkg);
        $res->[0] == 200 or return [
            500, "Can't get meta accessor for $pkg: $res->[0] - $res->[1]"];
        my $ma   = $res->[2];
        my $leaf = $f; $leaf =~ s/.+:://;
        my $meta = $ma->get_meta($pkg, $leaf)
            or return [500, "Can't get metadata for function $f"];
        my $ff  = $meta->{features} // {};
        my $ftx = $ff->{tx} && ($ff->{tx}{use} || $ff->{tx}{req});
        unless (
            ($ftx && $ff->{undo} && $ff->{idempotent}) ||
                $ff->{pure} ||
                    ($ff->{dry_run} && $cargs->{-dry_run})) {
            return [412, "Function not eligible for transaction: $f"];
        }
        if (defined $undo_data) {
            ref($undo_data) eq 'ARRAY'
                or return [400, "Undo data must be array"];
        }
        return [400, "Please specify undo data (step->[3])"]
            if $whichs eq 'call_undo' && !$undo_data;
    }

    # fargs = arguments to call our function with
    my %fargs = %{$fargs0 // {}};
    if ($whichs eq 'call') {
        if ($undo_data) {
            $fargs{-undo_action} = 'undo';
            $fargs{-undo_data}   = $undo_data;
        } else {
            $fargs{-undo_action} = 'do';
        }
    } else {
        $fargs{-undo_action} = 'redo';
        $fargs{-undo_data}   = $undo_data;
    }
    if ($whicha eq 'check') {
        $fargs{-dry_run} = 1;
        $fargs{-log_fix} = 0;
    } else {
        $fargs{-tx_manager} = $cargs->{-tx_manager};
        $fargs{-tx_call_id} = $cargs->{-tx_call_id};
    }

    no strict 'refs';
    my $res = *{$f}{CODE}->(%fargs);

    if ($whicha eq 'check') {
        if ($res->[0] == 200 || $res->[0] == 304) {
            return [200, "OK", [$us, $f, $fargs0, $res->[3]{undo_data} // []]];
        } else {
            return [500, "$f returns failure: $res->[0] - $res->[1]"];
        }
    } else {
        return $res;
    }
}

sub spec {
    state $step = {
        summary => 'Call another undoable local function',
        description => <<'_',

Syntax: `["call", $f, $args, $undo_data]`.

$f is the fully qualified function name (e.g. `Foo::Bar::func`), $args (hashref)
is the arguments hashref (defaults to {}), and $undo_data (arrayref, optional)
is undo data from function.

During `check` phase, will check the function's metadata to see if function is
indeed eligible (undoable, transactional, supports dry run, and is idempotent)
and the call $f under dry_run mode, to get function's undo data. Will return a
`call_undo` undo step: `["call_undo", $f, $args, $undo_data]`.

During `fix` phase, will call the function (passing -tx_manager, -undo_data, et
al) and return its result.

_
        check_or_fix => sub {
            __check_or_fix_for_call_or_call_undo('call', @_);
        },
    };
    $step;
}

1;
# ABSTRACT: Call another undoable local function

__END__
=pod

=head1 NAME

Perinci::Sub::Step::Common::call - Call another undoable local function

=head1 VERSION

version 0.23

=head1 DESCRIPTION


This module has L<Rinci> metadata.

=head1 FUNCTIONS


None are exported by default, but they are exportable.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Perinci/Sub/Gen/Undoable.pm
package Perinci::Sub::Gen::Undoable;

use 5.010;
use strict;
use warnings;
use Log::Any '$log';

use File::Spec;
use Perinci::Exporter;
use Perinci::Sub::Gen;
use Perinci::Sub::Wrapper qw(caller);
use Scalar::Util qw(blessed reftype);
use SHARYANTO::Log::Util qw(@log_levels);
use SHARYANTO::Package::Util qw(package_exists);

our $VERSION = '0.25'; # VERSION

our %SPEC;

$SPEC{gen_undoable_func} = {
    v           => 1.1,
    summary     => 'Generate undoable '.
        '(transactional, dry-runnable, idempotent) function',
    description => <<'_',

This function (hereby called generator, or helper) is a helper for writing
undoable (as well as transactional, dry-runnable, and idempotent) function
(hereby called generated function). Undo and transaction features are as
described in Rinci::function::Undo and Rinci::function::Transaction. Generator
will create the generated function with the basic flow and structure; you just
need to supply the necessary "business logics" bits like the `check_state` and
`fix_state` hooks.

Here is the flow of the generated function:

1. Execute `check_args` hook if supplied. Hook will be passed ($args). The task
   of the this hook is to check arguments (and possibly supply defaults). It
   should return enveloped result. If response is not success, exit with this
   response. Otherwise, continue.

2. If we are called in do mode (-undo_action => 'do' or undefined):

   2a. Execute `check_state` hook, to check state. Hook will be passed ($args).
       Hook must return enveloped result, status 304 if checking succeeds and
       fixed state (final, desired state) has been reached so nothing needs to
       be done, status 200 if checking succeeds and we are in a fixable state (a
       non-final state but fixable into the final state, also called an
       acceptable initial state), status 412 if checking succeeds but we are in
       an unfixable state (non-final state which is not fixable into the final
       state, also called an unacceptable initial state). Other status means
       checking has failed. Function should return with this result unless
       status is 200.

       If status is 200, hook must also return undo data (a list of calls that
       can be used to *undo* the 'fix_state' routine back to the initial state)
       in its result metadata, in `undo_data` key. This means, if we fail during
       fixing, we can rollback to initial state using this information.

       The undo data looks like this: C<[[f1, args1], [f2, args2], ...]>. C<fN>
       is the fully-qualified function name to call (required), C<argsN> is a
       hashref containing arguments for the function (optional, defaults to {}).

   2b. If function is called in dry run mode (-dry_run => 1), we return 200
       (along undo data from step 2a in the result metadata).

   2c. We execute `fix_state` hook to fix state to final/fixed state. Hook will
       be passed ($args, $undo_data), where $undo_data was retrieved from
       `check_state` hook. Hook must return enveloped result. Function returns
       with this result. Failure will trigger a rollback by TM.

3. If we are called in undo mode (-undo_action => 'undo', -undo_data => $calls):
   then we simply invoke TM's call() to process the undo calls.

For examples, see Setup::* Perl modules.

_
    args => {
        v => {
            summary => 'Interface version',
            schema => [int => {default=>1}],
            description => <<'_',

When Perinci::Sub::Gen::Undoable 0.22 broke backward-compatibility, v was
introduced at v=2 (previous interface at v=1).

For new interface, must be set to 2.

_
        },
        %Perinci::Sub::Gen::common_args,
        trash_dir => {
            summary => 'Whether function needs undo trash directory',
            schema  => [bool => {default => 0}],
            description => <<'_',

If set to true, generator will put `-undo_trash_dir` in function arguments.

_
        },
        args => {
            summary => 'Specification for generated function\'s arguments',
            schema  => [hash => {default=>{}}],
            description => <<'_',

This is just like the metadata property 'args'.

_
        },
        check_args => {
            summary => 'Code to check function\'s arguments',
            schema  => 'code*',
            description => <<'_',

See generator's documentation on how this hook should do and return.

_
        },
        check_state => {
            summary => 'Code to check state',
            schema  => 'code*',
            req     => 1,
            description => <<'_',

See generator's documentation on how this hook should do and return.

_
        },
        fix_state => {
            summary => 'Code to fix state',
            schema  => 'code*',
            description => <<'_',

See generator's documentation on how this hook should do and return.

_
        },
        check_or_fix_state => {
            summary => 'Code to check or fix state',
            schema  => 'code*',
            description => <<'_',

An alternative to combine check_state and fix_state in one code. Code will be
called with either ('check', $args) or ('fix', $args, $undo_data), depending on
whether check/fix is wanted.

_
        },
    } # args,
};
sub gen_undoable_func {
    my %gen_args = @_;
    my @caller = caller();

    my $v = $gen_args{v} // 1;
    if ($v == 1) {
        return[412, join(
            "",
            "$caller[0] is using old interface (v=1) of ", __PACKAGE__,
                " (v0.21 and lower). Current interface version is v=2. ",
                    "Please upgrade $caller[0] or use older version of ",
            __PACKAGE__)];
    }

    # XXX schema
    my ($uqname, $package);
    my $fqname = $gen_args{name};
    return [400, "Please specify name"] unless $fqname;
    if ($fqname =~ /(.+)::(.+)/) {
        $package = $1;
        $uqname  = $2;
    } else {
        $package = $gen_args{package} // $caller[0];
        $uqname  = $fqname;
        $fqname  = "$package\::$uqname";
    }
    return [400,"Please specify check_state&fix_state (or check_or_fix_state)"]
        unless $gen_args{check_state} && $gen_args{fix_state} ||
            $gen_args{check_or_fix_state};

    my $meta = {
        v           => 1.1,
        summary     => $gen_args{summary},
        description => $gen_args{description},
        args        => $gen_args{args} // {},
        features    => {
            undo       => 1,
            dry_run    => 1,
            idempotent => 1,
            tx         => {use=>1, req=>1},
        },
        deps        => {},
    };
    $meta->{deps}{undo_trash_dir} = 1 if $gen_args{trash_dir};

    my $code = sub {
        my %fargs = @_;
        my $res;

        if ($gen_args{check_args}) {
            $res = $gen_args{check_args}->(\%fargs);
            return [400, "Error in arguments: $res->[1]"]
                unless $res->[0] == 200;
        }

        my $dry_run     = $fargs{-dry_run};
        my $undo_action = $fargs{-undo_action} // "";
        return [400, "Invalid -undo_action, please use do/undo/redo only"]
            unless !length($undo_action) || $undo_action =~ /\A(un|re)?do\z/;
        my $tm          = $fargs{-tx_manager} or
            return [412, "This function requires transaction, ".
                        "please specify -tx_manager"];

        if ($gen_args{trash_dir}) {
            $res = $tm->get_trash_dir;
            return [500, "Can't get trash dir: $res->[0] - $res->[1]"]
                unless $res->[0] == 200;
            $fargs{-undo_trash_dir} //= $res->[2];
            # XXX this actually has a side-effect of creating trash_dir even
            # under dry_run. is this ok?
        }

        my $calls;
        my $is_rollback;

        if ($undo_action eq 'do' || !$undo_action) {

            # under do mode, we execute check_state then fix_state
            my $cof = $gen_args{check_or_fix_state};
            if ($cof) {
                $res = $cof->('check', \%fargs);
            } else {
                $res = $gen_args{check_state}->(\%fargs);
            }
            $log->tracef("Result of check_state: %s", $res)
                if $fargs{-log_call} // 1;
            return $res unless $res->[0] == 200;
            return $res if $dry_run;

            my $ud = $res->[3]{undo_data};
            $log->tracef("Fixing state ...");
            if ($cof) {
                $res = $cof->('fix', \%fargs, $ud);
            } else {
                $res = $gen_args{fix_state}->(\%fargs, $ud);
            }
            $log->tracef("Result of fix_state: %s", $res);
            if ($res->[0] == 200 || $res->[0] == 304) {
                $res->[3]{undo_data} //= $ud;
            }

            return $res; # fail, let TM do the roll back
        }

        # when in undo/redo mode, we need to perform a list of calls. note that
        # currently we don't load the function's module, wrap, get the
        # function's metadata, check function's transaction support, etc. if you
        # need all of those to be safer, use TM. our own loop is just the
        # crutch.

        $calls = $fargs{-undo_data}
            or return [400, "Please supply -undo_data for undo/redo"];
        $log->tracef("Undo calls to perform: %s", $calls);
        return $tm->call(calls => $calls, dry_run=>$dry_run);

    };

    if ($gen_args{install} // 1) {
        no strict 'refs';
        *{ $fqname } = $code;
        ${$package . "::SPEC"}{$uqname} = $meta; # XXX use Perinci::MetaAccessor
    }

    [200, "OK", {code=>$code, meta=>$meta}];
}

1;
# ABSTRACT: Generate undoable (transactional, dry-runnable, idempotent) function


__END__
=pod

=head1 NAME

Perinci::Sub::Gen::Undoable - Generate undoable (transactional, dry-runnable, idempotent) function

=head1 VERSION

version 0.25

=head1 SYNOPSIS

 # See an example, like in Setup::File::Symlink

=head1 DESCRIPTION

This module helps you write undoable/transactional functions (as well as
functions that support dry-run and are idempotent).

=head1 SEE ALSO

L<Rinci::function::Undo>, L<Rinci::function::Transaction>, L<Rinci>

L<Setup>.

=head1 DESCRIPTION


This module has L<Rinci> metadata.

=head1 FUNCTIONS


None are exported by default, but they are exportable.

=head2 gen_undoable_func(%args) -> [status, msg, result, meta]

Generate undoable (transactional, dry-runnable, idempotent) function.

This function (hereby called generator, or helper) is a helper for writing
undoable (as well as transactional, dry-runnable, and idempotent) function
(hereby called generated function). Undo and transaction features are as
described in Rinci::function::Undo and Rinci::function::Transaction. Generator
will create the generated function with the basic flow and structure; you just
need to supply the necessary "business logics" bits like the C<check_state> and
C<fix_state> hooks.

Here is the flow of the generated function:

=over

=item 1.

Execute C<check_args> hook if supplied. Hook will be passed ($args). The task
   of the this hook is to check arguments (and possibly supply defaults). It
   should return enveloped result. If response is not success, exit with this
   response. Otherwise, continue.



=item 1.

If we are called in do mode (-undo_action => 'do' or undefined):

   2a. Execute C<check_state> hook, to check state. Hook will be passed ($args).
   Hook must return enveloped result, status 304 if checking succeeds and
   fixed state (final, desired state) has been reached so nothing needs to
   be done, status 200 if checking succeeds and we are in a fixable state (a
   non-final state but fixable into the final state, also called an
   acceptable initial state), status 412 if checking succeeds but we are in
   an unfixable state (non-final state which is not fixable into the final
   state, also called an unacceptable initial state). Other status means
   checking has failed. Function should return with this result unless
   status is 200.

   If status is 200, hook must also return undo data (a list of calls that
   can be used to I<undo> the 'fix_state' routine back to the initial state)
   in its result metadata, in C<undo_data> key. This means, if we fail during
   fixing, we can rollback to initial state using this information.

   The undo data looks like this: C. C
   is the fully-qualified function name to call (required), C is a
   hashref containing arguments for the function (optional, defaults to {}).

   2b. If function is called in dry run mode (-dry_run => 1), we return 200
   (along undo data from step 2a in the result metadata).

   2c. We execute C<fix_state> hook to fix state to final/fixed state. Hook will
   be passed ($args, $undoI<data), where $undo>data was retrieved from
   C<check_state> hook. Hook must return enveloped result. Function returns
   with this result. Failure will trigger a rollback by TM.



=item 1.

If we are called in undo mode (-undoI<action => 'undo', -undo>data => $calls):
   then we simply invoke TM's call() to process the undo calls.



=back

For examples, see Setup::* Perl modules.

Arguments ('*' denotes required arguments):

=over 4

=item * B<args> => I<hash> (default: {})

Specification for generated function's arguments.

This is just like the metadata property 'args'.

=item * B<check_args> => I<code>

Code to check function's arguments.

See generator's documentation on how this hook should do and return.

=item * B<check_or_fix_state> => I<code>

Code to check or fix state.

An alternative to combine checkI<state and fix>state in one code. Code will be
called with either ('check', $args) or ('fix', $args, $undo_data), depending on
whether check/fix is wanted.

=item * B<check_state>* => I<code>

Code to check state.

See generator's documentation on how this hook should do and return.

=item * B<description> => I<str>

Generated function's description.

=item * B<fix_state> => I<code>

Code to fix state.

See generator's documentation on how this hook should do and return.

=item * B<install> => I<bool> (default: 1)

Whether to install generated function (and metadata).

By default, generated function will be installed to the specified (or caller's)
package, as well as its generated metadata into %SPEC. Set this argument to
false to skip installing.

=item * B<name>* => I<str>

Generated function's name, e.g. `myfunc`.

=item * B<package> => I<str>

Generated function's package, e.g. `My::Package`.

This is needed mostly for installing the function. You usually don't need to
supply this if you set C<install> to false.

If not specified, caller's package will be used by default.

=item * B<summary> => I<str>

Generated function's summary.

=item * B<trash_dir> => I<bool> (default: 0)

Whether function needs undo trash directory.

If set to true, generator will put C<-undo_trash_dir> in function arguments.

=item * B<v> => I<int> (default: 1)

Interface version.

When Perinci::Sub::Gen::Undoable 0.22 broke backward-compatibility, v was
introduced at v=2 (previous interface at v=1).

For new interface, must be set to 2.

=back

Return value:

Returns an enveloped result (an array). First element (status) is an integer containing HTTP status code (200 means OK, 4xx caller error, 5xx function error). Second element (msg) is a string containing error message, or 'OK' if status is 200. Third element (result) is optional, the actual result. Fourth element (meta) is called result metadata and is optional, a hash that contains extra information.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

